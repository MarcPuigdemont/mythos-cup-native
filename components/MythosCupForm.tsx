import React, { useState, useCallback } from 'react';
import { StyleSheet, View } from 'react-native';
import { Input, Button, Text } from 'react-native-elements';
import { useDispatch } from 'redux-react-hook';

import { ICup } from '../interfaces';

import CreateMythosCupIcons from '../components/CreateMythosCupIcons';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignSelf: "stretch",
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
    padding: 20,
    backgroundColor: '#fff',
  },
  input: {
    marginBottom: 30,
  },
  text: {
    fontSize: 16,
    fontFamily: 'sans-serif',
    fontWeight: 'bold',
    color: '#86939e',
    alignSelf: "flex-start",
    marginLeft: 10,
    marginBottom: 20,
  },
});

interface Props {
  navigation: { navigate: Function };
  cup: ICup;
  action: Function;
};
const MythosCupForm = (props: Props) => {
  const { navigate } = props.navigation;
  const [ campaign, setCampaign ] = useState(props.cup ? props.cup.campaign : '');
  const [ difficulty, setDifficulty ] = useState(props.cup ? props.cup.difficulty : '');
  const [ icon, selectIcon ] = useState(props.cup ? props.cup.icon : 0);
  const dispatch = useDispatch();
  const handleAddCup = useCallback(() => {
    const cup = Object.assign({}, props.cup || {}, { campaign, difficulty, icon });
    dispatch(props.action(cup));
    setCampaign('');
    setDifficulty('');
    selectIcon(0);
    navigate('MythosCupsList');
  }, [campaign, difficulty, icon]);
  return (
    <View style={styles.container}>
      <Input containerStyle={styles.input} label='Which scenario are you playing' onChangeText={setCampaign} value={campaign} />
      <Input containerStyle={styles.input} label='Which difficulty are you playing on' onChangeText={setDifficulty} value={difficulty} />
      <Text style={styles.text}>Give it an icon to represent the campaign</Text>
      <CreateMythosCupIcons selected={icon} onSelect={selectIcon} />
      <Button title={ props.cup ? "Edit" : "Create"} onPress={handleAddCup} />
    </View>
  );
};

export default MythosCupForm;