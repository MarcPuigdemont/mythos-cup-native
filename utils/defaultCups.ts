import { ICup } from "../interfaces";

const defaultCups: ICup[] = [
  {
    id: '1',
    campaign: 'Fest for Umorhoth',
    difficulty: 'Easy',
    icon: 2,
    tokens: null,
    revealedTokens: null,
    unRevealedTokens: null,
    playTokens: null,
  },
  {
    id: '2',
    campaign: 'Fest for Umorhoth',
    difficulty: 'Not that hard',
    icon: 4,
    tokens: null,
    revealedTokens: null,
    unRevealedTokens: null,
    playTokens: null,
  },
  {
    id: '3',
    campaign: 'Whispers in the dark',
    difficulty: 'Hard',
    icon: 7,
    tokens: null,
    revealedTokens: null,
    unRevealedTokens: null,
    playTokens: null,
  },
];

export default defaultCups;