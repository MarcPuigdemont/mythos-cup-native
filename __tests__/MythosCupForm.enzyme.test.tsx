/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import MythosCupForm from '../components/MythosCupForm';
import defaultCups from '../utils/defaultCups';
import { act } from 'react-dom/test-utils';

const cup = defaultCups[0];
let mockNavigate = jest.fn();
const navigation = { navigate: mockNavigate };
let mockDispatch = jest.fn();
jest.mock('redux-react-hook', () => {
    return {
      useDispatch: () => mockDispatch
    };
});
jest.mock('../components/CreateMythosCupIcons', () => {
  const CreateMythosCupIconsMock = () => {
    return <div id="CreateMythosCupIcons"></div>;
  }
  return CreateMythosCupIconsMock;
});

describe('MythosCupForm', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
        <MythosCupForm cup={cup} navigation={navigation} action={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();
    
  });
  it('should dispatch a valid cup and navigate', () => {
    mockDispatch = jest.fn();
    const action = jest.fn().mockImplementation((a) => a);
    const wrapper = mount(
        <MythosCupForm cup={null} navigation={navigation} action={action} />
    );
    expect(wrapper.find('Input')).toHaveLength(2);
    expect(wrapper.find('CreateMythosCupIconsMock')).toHaveLength(1);
    expect(wrapper.find('Button')).toHaveLength(1);

    act(() => {
      const inputValues = [ 'campaign 1', 'easy'];
      wrapper.find('Input').forEach((el, i) => el.props().onChangeText(inputValues[i]));
      wrapper.find('CreateMythosCupIconsMock').props().onSelect(4);
    });
    wrapper.update();
    act(() => {
      wrapper.find('Button').props().onPress();
    });
    wrapper.update();

    expect(action).toHaveBeenCalledWith({ campaign: 'campaign 1', difficulty: 'easy', icon: 4 });
    expect(mockDispatch).toHaveBeenCalledWith({ campaign: 'campaign 1', difficulty: 'easy', icon: 4 });
    expect(mockNavigate).toHaveBeenCalledWith('MythosCupsList');

    expect(wrapper.find('Input').map((el) => el.props().value)).toEqual(['', '']);
    expect(wrapper.find('CreateMythosCupIconsMock').props().selected).toEqual(0);
  });
})