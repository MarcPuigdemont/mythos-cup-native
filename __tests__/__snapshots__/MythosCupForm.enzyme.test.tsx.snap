// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MythosCupForm renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <ForwardRef(Themed.Input)
    containerStyle={
      Object {
        "marginBottom": 30,
      }
    }
    label="Which scenario are you playing"
    onChangeText={[Function]}
    value="Fest for Umorhoth"
  />
  <ForwardRef(Themed.Input)
    containerStyle={
      Object {
        "marginBottom": 30,
      }
    }
    label="Which difficulty are you playing on"
    onChangeText={[Function]}
    value="Easy"
  />
  <Themed.Text
    style={
      Object {
        "alignSelf": "flex-start",
        "color": "#86939e",
        "fontFamily": "sans-serif",
        "fontSize": 16,
        "fontWeight": "bold",
        "marginBottom": 20,
        "marginLeft": 10,
      }
    }
  >
    Give it an icon to represent the campaign
  </Themed.Text>
  <CreateMythosCupIconsMock
    onSelect={[Function]}
    selected={2}
  />
  <ForwardRef(Themed.Button)
    onPress={[Function]}
    title="Edit"
  />
</View>
`;
